query/sources/impl/src/main/java/org/codice/ddf/admin/sources/commons/utils/WfsSourceUtils.java

```diff
@@ -0,0 +1,160 @@
+                .map(url -> {
+                    UrlField urlField = new UrlField(hostField.fieldName());
+                    urlField.updatePath(hostField.path());
+                    urlField.setValue(url);
+                    return urlField;
+                })
+                .filter(urlField -> !sendWfsCapabilitiesRequest(urlField, creds).containsErrorMsgs())
+                .map(ReportWithResult::new)
+                .findFirst()
+                .orElse(createDefaultResult(hostField));
+    }
+
+    /**
+     * Attempts to create a WFS configuration from the given url.
+     *
+     * @param urlField WFS URL to probe for a configuration
+     * @param creds    optional username to add to Basic Auth header
+     * @return a {@link ReportWithResult} containing the preferred {@link SourceConfigUnionField}, or containing {@link org.codice.ddf.admin.common.message.ErrorMessage}s on failure.
+     */
+    public ReportWithResult<SourceConfigUnionField> getPreferredWfsConfig(UrlField urlField, CredentialsField creds) {
+        ReportWithResult<String> responseBodyResult = requestUtils.sendGetRequest(urlField, creds, GET_CAPABILITIES_PARAMS);
+        ReportWithResult<SourceConfigUnionField> configResult = new ReportWithResult<>();
+        if (responseBodyResult.containsErrorMsgs()) {
+            configResult.addMessages(responseBodyResult);
+            return configResult;
+
+        }
+
+        String requestBody = responseBodyResult.result();
+        Document capabilitiesXml;
+        try {
+            capabilitiesXml = createDocument(requestBody);
+        } catch (Exception e) {
+            LOGGER.debug("Failed to read response from WFS endpoint.");
+            configResult.argumentMessage(unknownEndpointError(urlField.path()));
+            return configResult;
+        }
+
+        WfsSourceConfigurationField preferredConfig = new WfsSourceConfigurationField();
+        preferredConfig.endpointUrl(urlField.getValue())
+                .credentials()
+                .username(creds.username())
+                .password(creds.password());
+
+        XPath xpath = XPathFactory.newInstance()
+                .newXPath();
+        xpath.setNamespaceContext(SOURCES_NAMESPACE_CONTEXT);
+        String wfsVersion;
+        try {
+            wfsVersion = xpath.compile(WFS_VERSION_EXP)
+                    .evaluate(capabilitiesXml);
+        } catch (XPathExpressionException e) {
+            LOGGER.debug("Failed to parse XML response.");
+            configResult.argumentMessage(unknownEndpointError(urlField.path()));
+            return configResult;
+        }
+
+        try {
+            configResult.result(preferredConfig.wfsVersion(wfsVersion));
+        } catch (IllegalArgumentException e) {
+            configResult.argumentMessage(unknownEndpointError(urlField.path()));
+        }
+        return configResult;
+    }
+
+    private ReportWithResult<UrlField> createDefaultResult(HostField hostField) {
```

Comment:
I don't think this method is necessary, anyone calling this can call `new ReportWithResult<UrlField>.argumentMessage(unknownEndpointError(hostField.path()));`